### PR TITLE
Issue 11 convert document to graph

### LINKED ISSUE #11: convert spdx document to graph

### PR DESCRIPTION
**Pull request by meretp**
This PR adds some basic logic to convert a parsed SPDX document to a directed graph. I decided to use the library `networkx` to build a graph as it is the most common and used in Python and seems to provide all functionality that we need at this point of development.

I started with a conversion method for SPDXLite files which only contain packages and have no relationships. As all packages are described by the document, I added edges with the name `DESCRIBES` for each package. 

The method `generate_graph` taks a complete spdx document as input and generates a graph according to the existing relationships. 

So far each node gets the attribute `element` which stores the complete object with all fields/ properties. We will need to find out how to represent all properties within the `.opossum`  file.

An open question is: How do we handle `NOASSERTION` or `NONE` in relationships? The current implementation will add a node at the first occurence of `NONE` or `NOASSERTION` and link all elements  that have a relationship to `NONE` or `NOASSERTION` with this node. 

I also added tests and test files (2 self-generated, one from the `tools-python` repo) to check that the generated graphs contain nodes for each element and edges for each relationship.

Edit: We agreed to add nodes for each relationship. To be able to collect all relationships for one source node and not use the same node for every relationship of one type I added the source node as prefix to the relationships node. This PR doesn't handle `Snippets` yet. Also we only add elements to the graph which are part of any relationship.  Otherwise we wouldn't be able to find a minimum spanning arborescence which we need to do in the next step.

fixes #11

### CODE REVIEW

File: tests/test_graph_generation.py
Comment (JakobSchubert) [inline tests/test_graph_generation.py:1]: We don't have any test that actually asserts the whole graph structure. Currently we e. g. wouldn't catch bugs related to the actual graph topology like all edges are starting at a fixed destination node etc. .
I would be fine with a very minimal example with 4 nodes or so but having such a test might help us in the future.

**Review by JakobSchubert — APPROVED** [review]
Looks great. Some small comments from my side. Also note that we might to remove the "test_hello_world" now as we have the first real tests.

File: src/opossum_lib/graph_generation.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+def generate_graph_from_spdx_lite(document: Document) -> DiGraph:
+    graph = DiGraph()
+
+    document_spdx_id = document.creation_info.spdx_id
+    graph.add_node(document_spdx_id)
+
+    graph.add_node("DESCRIBES")
+    graph.add_edge(document_spdx_id, "DESCRIBES")
+
+    package_spdx_ids = [package.spdx_id for package in document.packages]
```
[the remark this answers is not in the document]
Comment (meretp) [reply src/opossum_lib/graph_generation.py:20]: Correct, the SPDX identifiers uniquely identify any element in the document. 
References for that:
- https://spdx.github.io/spdx-spec/v2.3/snippet-information/#911-description
- https://spdx.github.io/spdx-spec/v2.3/file-information/#821-description
- https://spdx.github.io/spdx-spec/v2.3/package-information/#721-description
- https://spdx.github.io/spdx-spec/v2.3/document-creation-information/#63-spdx-identifier-field

File: src/opossum_lib/graph_generation.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    graph.add_nodes_from(package_spdx_ids)
+    graph.add_edges_from(("DESCRIBES", spdx_id) for spdx_id in package_spdx_ids)
+
+    return graph
+
+
+def generate_graph_from_file(document: Document) -> DiGraph:
```
[the remark this answers is not in the document]
Comment (meretp) [reply src/opossum_lib/graph_generation.py:27]: Done

File: src/opossum_lib/graph_generation.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            spdx_element_id = str(relationship.related_spdx_element_id)
+
+            if spdx_element_id not in graph.nodes():
+                graph.add_node(spdx_element_id)
+            graph.add_edge(relationship_node_key, spdx_element_id)
+
+    return graph
+
+
+def get_relationships_by_spdx_id(
```
[the remark this answers is not in the document]
Comment (meretp) [reply src/opossum_lib/graph_generation.py]: Yes, we also thought about this in the `tools-python` repo but didn't used it so far. I think this is a nice way to indicate helper functions.

File: src/opossum_lib/graph_generation.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+def generate_graph_from_file(document: Document) -> DiGraph:
+    graph = DiGraph()
+    graph.add_node(document.creation_info.spdx_id, element=document.creation_info)
+
+    relationships_by_spdx_id: Dict[
+        str, List[Relationship]
+    ] = get_relationships_by_spdx_id(document.relationships)
+
+    for spdx_id in relationships_by_spdx_id.keys():
```
[the remark this answers is not in the document]
Comment (meretp) [reply src/opossum_lib/graph_generation.py]: Good point.

File: src/opossum_lib/graph_generation.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        if spdx_id not in graph.nodes():
+            graph.add_node(spdx_id)
```
[the remark this answers is not in the document]
Comment (meretp) [reply src/opossum_lib/graph_generation.py]: Yes, I forgot it in some way. When I implemented this, we didn't have a helper method in `tools-python` to extract an element from the document by a given spdx_id. In the meantime, this functionality was added and I can now use this new helper method.

File: src/opossum_lib/graph_generation.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        for relationship in relationships_by_spdx_id[spdx_id]:
+            relationship_node_key = (
+                str(relationship.spdx_element_id)
```
[the remark this answers is not in the document]
Comment (meretp) [reply src/opossum_lib/graph_generation.py]: Correct, I mixed it with `relationship.related_spdx_element_id` which could also be `SpdxNone` or `SpdxNoAssertion`.

File: tests/test_graph_generation.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+from opossum_lib.graph_generation import (
+    generate_graph_from_file,
+    generate_graph_from_spdx_lite,
+)
+
+
+def test_generate_graph_from_spdx_lite() -> None:
+    document = parse_file(
+        os.path.join(os.path.dirname(__file__), "./data/SPDXLite.spdx")
```
[the remark this answers is not in the document]
Comment (meretp) [reply tests/test_graph_generation.py]: I didn't know `pathlib` before but a quick research about it convinced me and I will use it here.

File: tests/test_graph_generation.py
[the remark this answers is not in the document]
Comment (meretp) [reply tests/test_graph_generation.py:1]: With the assertion on the number of nodes and edges this is somehow done in the current implementation. We don't explicitly check a specific edge (or string like "SPDXRef-DOCUMENT_DESCRIBES") but if the concatenation wouldn't work as expected the number of nodes and edges would be less. But I admit that this is not really stable. As every generated graph should have a SPDXRef-DOCUMENT describes relationship I will add a few assertions for the relationship_node_keys.

Comment (meretp) [reply tests/test_graph_generation.py:1]: I am not really sure what to test about the graph topology at this point. For general SPDX documents the graph doesn't need to be connected. A document needs to describe at least one element so we could assert on at least one connected component of the graph that has a node `SPDXRef-DOCUMENT` (exactly this ID is required for the document) and at least one outgoing edge. But we can have other subgraphs that are not connected with this component. We can also have an element, say `SPDXRef-Package` that has a relationship that points to the document, so an ingoing edge to the `SPDXRef-DOCUMENT` node. If in this case there is no directed path from `SPDXRef-DOCUMENT` to `SPDXRef-Package` this  will lead to problems when generating the tree but I think this is a problem for later. So, coming back to the initial comment: Would a test for this one connected component with at least one document node and at least one outgoing edge be sufficient or do you have more in mind?

Comment (meretp) [reply tests/test_graph_generation.py:1]: Update: It seems that not even this minimal requirement is true. We are currently clarifying if every SPDX document needs to have at least one describes relationship or not, the spec is not really clear at this point. To continue here anyway, I would open an issue to add a more general test for the graph topology and address it once the confusion about the "DESCRIBES"-relationship is resolved.

**Comment by meretp** [comment]
While working on this, I had a deeper look into the specification of SPDXLite documents and it seems that they only allow one package. As this doesn't really make SPDXLite a good minimal example any more, I decided to delete this part and concentrate on the implemenation for general SPDX documents.

**Review by JakobSchubert — APPROVED** [review]
Awesome, thanks for implementing all the comments! I think we can merge this PR now.

**Pull request merged**